### PR TITLE
KAFKA-15830: Add telemetry API handling (KIP-714)

### PR DESCRIPTION
**Pull request by apoorvmittal10**
The PR adds handling of telemetry APIs in KafkaAPIs.scala which calls the respective manager to handle the API calls. Also the telemetry plugin which if registered in configs get registered for exporting client metrics.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/java/kafka/metrics/ClientMetricsReceiverPlugin.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/metrics/ClientMetricsReceiverPlugin.java]: This class is major overlap with the other open PR: https://github.com/apache/kafka/pull/14699.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -3697,16 +3698,60 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-  // Just a place holder for now.
   def handleGetTelemetrySubscriptionsRequest(request: RequestChannel.Request): Unit = {
-    requestHelper.sendMaybeThrottle(request, request.body[GetTelemetrySubscriptionsRequest].getErrorResponse(Errors.UNSUPPORTED_VERSION.exception))
-    CompletableFuture.completedFuture[Unit](())
+    val subscriptionRequest = request.body[GetTelemetrySubscriptionsRequest]
+
+    if (ClientMetricsReceiverPlugin.instance.isEmpty) {
+      info("Received get telemetry client request, no metrics receiver plugin configured or running with ZK")
+      requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs =>
+        subscriptionRequest.getErrorResponse(requestThrottleMs, Errors.UNSUPPORTED_VERSION.exception))
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: @AndrewJSchofield Though the KIP-714 says that `APIVersions` response should mention telemetry APIs are not supported if ClientMetrics Exporter Plugin is not configured. But I didn't get a way to do that in APIVersions API as the API seems very generic which works with version number configured. I might be missing something here and would wait for @junrao's suggestion. For now I have added the check in Telemetry API itself.

**Comment by apoorvmittal10** [comment]
Build depends on PR: https://github.com/apache/kafka/pull/14699

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala [hunk 3]
```diff
@@ -913,6 +915,13 @@ class DynamicMetricReporterState(brokerId: Int, config: KafkaConfig, metrics: Me
     reporters.forEach { reporter =>
       metrics.addReporter(reporter)
       currentReporters += reporter.getClass.getName -> reporter
+      val clientTelemetryReceiver = reporter match {
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/DynamicBrokerConfig.scala]: Changes not required now, and the file shall be removed when merged with open PR: https://github.com/apache/kafka/pull/14699

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: We can wait for @junrao to weigh in, but I think the handling of `APIVersions` should be as written in the KIP. The whole point is that a client can tell from the `APIVersionsResponse` that there is no need to get a telemetry subscription. For a cluster without client telemetry enabled, there are no additional RPCs when the client connects. With this PR, a 3.7 client will make extra RPCs even when client telemetry is not enabled. This should be avoided if at all possible.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 4]
```diff
@@ -3747,16 +3747,56 @@ class KafkaApis(val requestChannel: RequestChannel,
...
   def handleGetTelemetrySubscriptionsRequest(request: RequestChannel.Request): Unit = {
-    requestHelper.sendMaybeThrottle(request, request.body[GetTelemetrySubscriptionsRequest].getErrorResponse(Errors.UNSUPPORTED_VERSION.exception))
-    CompletableFuture.completedFuture[Unit](())
+    val subscriptionRequest = request.body[GetTelemetrySubscriptionsRequest]
+
+    clientMetricsManager match {
+      case Some(metricsManager) =>
+        try {
+          if (metricsManager.isTelemetryReceiverConfigured) {
+            requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs =>
```
Comment (junrao) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: If we use `sendResponseMaybeThrottle`, we need to set requestThrottleMs in the response. A simpler approach is to use `sendMaybeThrottle`, which automatically sets `requestThrottleMs` in the response. Ditto in other usage.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (junrao) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yes, I agree that we need to exclude telemetry requests in `APIVersionsResponse` if ClientMetrics Exporter Plugin is not configured. We could do that by passing in the plugin availability flag through `ApiVersionManager.apiVersionResponse`.

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Thanks for the PR. Left a couple of comments.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @junrao Thanks for the suggestion. Just want to clarify, does adding `clientTelemetryEnabled` in `apiVersionResponse` requires another KIP or we can add that in KIP-714 or nothing to be documented?

Comment (junrao) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @apoorvmittal10 : ApiVersionResponse is defined to include all new types of requests supported on the server. So, there is no need to have a separate KIP. There is no need to add clientTelemetryEnabled in ApiVersionResponse. If clientTelemetryEnabled is false on the server, we can just exclude telemetry related API keys from ApiVersionResponse.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 4 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks @junrao , done.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks for the quick reply @junrao, I am looking into this. I thought the suggestion was to have a flag similar to `zkMigrationEnabled`, but it looks like to not emit APIKeys of telemetry if plugin is missing.

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Done.

**Comment by apoorvmittal10** [comment]
@junrao @AndrewJSchofield I have added the filtering for telemetry keys if plugin is not configured, please if the changes can be re-reviewed. I have addressed other feedback as well.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 5]
```diff
@@ -235,14 +235,15 @@ class BrokerServer(
...
+      clientMetricsManager = new ClientMetricsManager(clientMetricsReceiverPlugin, config.clientTelemetryMaxBytes, time)
 
       val apiVersionManager = ApiVersionManager(
         ListenerType.BROKER,
         config,
         Some(forwardingManager),
         brokerFeatures,
-        metadataCache
+        metadataCache,
+        Some(clientMetricsManager)
```
Comment (junrao) [inline core/src/main/scala/kafka/server/BrokerServer.scala:246]: We need to be a bit careful about synchronization. `ClientMetricsReceiverPlugin.receivers` is set in the main thread later through `config.dynamicConfig.addReconfigurables(this)`, but is read in either the network or the request handler thread. We need to add memory barrier to ensure visibility across threads.

File: core/src/main/scala/kafka/server/ApiVersionManager.scala [hunk 6]
```diff
@@ -130,7 +133,8 @@ class DefaultApiVersionManager(
   brokerFeatures: BrokerFeatures,
   metadataCache: MetadataCache,
   val enableUnstableLastVersion: Boolean,
-  val zkMigrationEnabled: Boolean = false
+  val zkMigrationEnabled: Boolean = false,
+  val clientMetricsManager: Option[ClientMetricsManager] = None
```
Comment (junrao) [inline core/src/main/scala/kafka/server/ApiVersionManager.scala:138]: Could we add the new param to javadoc?

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Thanks for the update PR. A couple of more comments.

File: core/src/main/scala/kafka/server/BrokerServer.scala (hunk 5 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/BrokerServer.scala:246]: Moved it synchronized list, done.

File: core/src/main/scala/kafka/server/ApiVersionManager.scala (hunk 6 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/ApiVersionManager.scala:138]: Missed that, my bad. Done.

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Thanks for the updated PR. The code LGTM. Waiting for the tests to finish.

**Comment by apoorvmittal10** [comment]
The CI seems to be unpredictable with randomly not able to checkout from HEAD for different JDK, not deterministic.

**Comment by junrao** [comment]
@apoorvmittal10 : Are the 59 test failures related, especially those ApiVersionsRequestTest?

**Comment by apoorvmittal10** [comment]
> @apoorvmittal10 : Are the 59 test failures related, especially those ApiVersionsRequestTest?

Hi @junrao, I should have realized that integration test in `ApiVersionsRequestTest` are legitimate before re-running the build. I have added fix in the PR where filtering needs to be applied while determining the keys returned as per config in integration tests.

Waiting for the build to finish, I ll check the tests.

**Comment by apoorvmittal10** [comment]
@junrao The build passed on all JDK with flaky tests failure.

There are some consistent failures as well which are unrelated to the PR changes.

**Comment by apoorvmittal10** [comment]
@junrao Below are the status of flaky tests. I have ran locally the flaky tests and can see them passed. I have triaged some of the failing tests, others have existing jira.

`Existing`: Build / JDK 17 and Scala 2.13 / testOffsetTranslationBehindReplicationFlow() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
40s: https://issues.apache.org/jira/browse/KAFKA-15197

`Triaged`: Build / JDK 17 and Scala 2.13 / testQuotaOverrideDelete(String).quorum=kraft – kafka.api.ClientIdQuotaTest
2s: https://issues.apache.org/jira/browse/KAFKA-15960

`Triaged`: Build / JDK 17 and Scala 2.13 / testTopicIdPersistsThroughControllerRestart() – kafka.controller.ControllerIntegrationTest
<1s: https://issues.apache.org/jira/browse/KAFKA-15961

`Triaged`: Build / JDK 17 and Scala 2.13 / testAlterPartitionVersion2KeepWorkingWhenControllerDowngradeToPre28IBP() – kafka.controller.ControllerIntegrationTest
<1s: https://issues.apache.org/jira/browse/KAFKA-15962

`Triaged`: Build / JDK 17 and Scala 2.13 / testBrokerHeartbeatDuringMigration [3] 3.6-IV0 – org.apache.kafka.controller.QuorumControllerTest
<1s: https://issues.apache.org/jira/browse/KAFKA-15963

`Triaged`: Build / JDK 17 and Scala 2.13 / testHighAvailabilityTaskAssignorLargeNumConsumers – org.apache.kafka.streams.processor.internals.StreamsAssignmentScaleTest
1m 35s: https://issues.apache.org/jira/browse/KAFKA-15964

`Existing`: Build / JDK 8 and Scala 2.12 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
1m 49s: https://issues.apache.org/jira/browse/KAFKA-15945

`Existing`: Build / JDK 8 and Scala 2.12 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
2m 5s: https://issues.apache.org/jira/browse/KAFKA-15927

`Existing`: Build / JDK 8 and Scala 2.12 / testWithGroupMetadata() – kafka.api.TransactionsBounceTest
28s: https://issues.apache.org/jira/browse/KAFKA-10251

`Existing`: Build / JDK 8 and Scala 2.12 / testWithGroupMetadata() – kafka.api.TransactionsBounceTest
21s: https://issues.apache.org/jira/browse/KAFKA-10251

`Existing`: Build / JDK 21 and Scala 2.13 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest: https://issues.apache.org/jira/browse/KAFKA-15675

`Existing`: Build / JDK 11 and Scala 2.13 / testWithGroupMetadata() – kafka.api.TransactionsBounceTest: https://issues.apache.org/jira/browse/KAFKA-10251

**Comment by junrao** [comment]
@apoorvmittal10 : Thanks for triaging the tests. There are a bunch of failures for BrokerRegistrationRequestTest. Is that being tracked?

**Comment by apoorvmittal10** [comment]
> @apoorvmittal10 : Thanks for triaging the tests. There are a bunch of failures for BrokerRegistrationRequestTest. Is that being tracked?

I have created jira to track that: https://issues.apache.org/jira/browse/KAFKA-15965

Can see jira already created for `ConsistencyVectorIntegrationTest`: https://issues.apache.org/jira/browse/KAFKA-15957

**Comment by apoorvmittal10** [comment]
@junrao Please let me know if I need to address anything else or the PR is good to merge.

**Comment by junrao** [comment]
@apoorvmittal10 : Now that https://github.com/apache/kafka/pull/14887 is merged. Triggering a rerun of the tests.

**Comment by apoorvmittal10** [comment]
> @apoorvmittal10 : Now that #14887 is merged. Triggering a rerun of the tests.

Thanks @junrao

**Comment by apoorvmittal10** [comment]
@junrao I think the merge from upstream is required to get the changes in current branch. I have merged upstream/trunk.

**Comment by junrao** [comment]
@apoorvmittal10 : Thanks for triaging the tests. Merging the PR.

**Pull request merged**